pkg/controller/backend/backend_controller.go

```diff
@@ -0,0 +1,223 @@
+	// Create a new controller
+	c, err := controller.New("backend-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource Backend
+	err = c.Watch(&source.Kind{Type: &capabilitiesv1beta1.Backend{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileBackend implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileBackend{}
+
+// ReconcileBackend reconciles a Backend object
+type ReconcileBackend struct {
+	*reconcilers.BaseReconciler
+}
+
+// Reconcile reads that state of the cluster for a Backend object and makes changes based on the state read
+// and what is in the Backend.Spec
+func (r *ReconcileBackend) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := r.Logger().WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconcile Backend", "Operator version", version.Version)
+
+	// Fetch the Backend instance
+	backend := &capabilitiesv1beta1.Backend{}
+	err := r.Client().Get(context.TODO(), request.NamespacedName, backend)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			reqLogger.Info("resource not found. Ignoring since object must have been deleted")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if reqLogger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(backend, "", "  ")
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		reqLogger.V(1).Info(string(jsonData))
+	}
+
+	// Ignore deleted Backends, this can happen when foregroundDeletion is enabled
+	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
+	if backend.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcile(backend)
+	if err != nil {
+		reqLogger.Error(err, "Failed to reconcile")
+		r.EventRecorder().Eventf(backend, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+	}
+	reqLogger.Info("END", "result", result, "error", err)
+	return result, err
+}
+
+func (r *ReconcileBackend) reconcile(backendResource *capabilitiesv1beta1.Backend) (reconcile.Result, error) {
+	logger := r.Logger().WithValues("reconcile", backendResource.Name)
+
+	if backendResource.SetDefaults() {
+		err := r.Client().Update(r.Context(), backendResource)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Failed setting backend defaults: %w", err)
+		}
+
+		logger.Info("resource defaults updated. Requeueing.")
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	backendAPI, specErr := r.reconcileSpec(backendResource)
```

Comment:
Is there never a scenario where we want to requeue due to some reason before reconciling the whole spec? For example because step B requires step A be reconciled and changes effective before running step B

Reply:
`reconcileSpec` tries to reconcile backend spec with 3scale control plane. Does not know anything about requeing or k8s stuff.

There are some errors it can raise that the backend controller will read them and requeue, but that logic belong to the backend controller.

Reply:
You might want to be able to requeue without an error between steps of the reconciler, which does not seem possible with this approach as the information that the backend controller receives is an error.
Might be worth to add that possibility in the future.